### PR TITLE
[Docs] Cross-reference concept exercise file information

### PR DESCRIPTION
**Pull request by ErikSchierboom**
This PR tries to apply DRY to the descriptions of the various Concept Exercises files. So far, we've duplicated the information in several files to make for easy reading for the reader (no hyperlinks to follow). However, keeping this information up to date has become a chore and things will (and have) be missed when updating them. This PR attempts to fix this by linking to the description in a single file (`concept-exercises.md`). Furthermore, it:

- Tries to be more specific on if a file is created or updated.
- Updates the order of the files to make more sense.
- Removes the step numbers, which imply an ordering that actually is not the optimal order (there is no such ordering).
- Add an example file when possible.
- Add missing information (e.g. the `config.json` file).

I've also updated the C# "implementing-.." document to a format that I think is more helpful.

### CODE REVIEW

File: docs/concept-exercises.md [hunk 1]
```diff
@@ -49,9 +49,9 @@ When designing Concept Exercises, please consider the following guidelines:
 
 When writing documentation and supporting Markdown files for Concept Exercises, please consult the [style guide][style-guide] noting the [automatic formatting section][style-guide-auto-formatting]. Also check any language-specific style guides, where applicable.
 
-## Exercise Structure
+## Exercise-specific files
```
Comment (iHiD) [inline docs/concept-exercises.md:52]: Suggestion that we link to the YouTube video here?

File: docs/maintainers/generic-how-to-implement-a-concept-exercise.md [hunk 2]
```diff
@@ -28,41 +28,75 @@ languages
...
+
+**Purpose:** Introduce the concept(s) that the exercise teaches to the student.
+
+For more information, please read [this in-depth description][docs-introduction.md] and check [this example file][example-docs-introduction.md].
+
+## Create `.docs/instructions.md` file
+
+**Purpose:** Provide instructions for the exercise.
+
+For more information, please read [this in-depth description][docs-instructions.md] and check [this example file][example-docs-instructions.md].
```
Comment (iHiD) [inline docs/maintainers/generic-how-to-implement-a-concept-exercise.md:53]: Link to YT here?

```suggestion
For more information, please read [this in-depth description][docs-instructions.md], watch [this video](...) and check [this example file][example-docs-instructions.md].
```

File: docs/concept-exercises.md (hunk 1 above)
Comment (ErikSchierboom) [reply docs/concept-exercises.md:52]: Updated.

File: docs/maintainers/generic-how-to-implement-a-concept-exercise.md (hunk 2 above)
Comment (ErikSchierboom) [reply docs/maintainers/generic-how-to-implement-a-concept-exercise.md]: Done.

**Comment by SaschaMann** [comment]
What's DRY?

**Comment by coriolinus** [comment]
@SaschaMann Don't Repeat Yourself: factor the code such that there are no cases of copy and paste.

File: languages/c/reference/implementing-a-concept-exercise.md [hunk 3]
```diff
@@ -44,17 +44,17 @@ languages
                 └── test-framework
 </pre>
 
-## Step 1: Adding Track-Specific Files
+## Step 1: adding code files
```
Comment (iHiD) [inline languages/c/reference/implementing-a-concept-exercise.md:47]:
```suggestion
## Step 1: Adding code files
```

File: languages/c/reference/implementing-a-concept-exercise.md [hunk 4]
```diff
@@ -44,17 +44,17 @@ languages
...
+The code files are track-specific and should be designed to help the student learn the exercise's concepts. The following C code files must be created (not necessarily in this order):
 
 - `makefile`: the makefile for the stub files and tests
 - `<NAME>{.c|.h}`: the stub source and header implementation files, which are the starting point for students to work on the exercise.
 - `test_<NAME>.c`: the unit test source file.
 - `test-framework`: this directory contains the unit test framework, [Unity][unity]
 - `.meta/example{.c|.h}`: an example implementation that passes all the unit tests.
 
-## Step 2: Adding Common Files
+## Step 2: adding documentation files
```
Comment (iHiD) [inline languages/c/reference/implementing-a-concept-exercise.md:57]:
```suggestion
## Step 2: Add documentation files
```

**Comment by verdammelt** [comment]
This looks good to me, I was wanting to de-dup the common-lisp 'implementing' doc from the shared one myself.

But should we go a little further and remove the non-track-specific files from all the track implementing docs? That is in the directory tree displayed in each one of these docs we list both track-specific and non-track-specific files. If in the future we want to add a new `meta.config` file (or some such) then every one of the tracks implementing docs needs to be updated to list that file it its directory tree.

I'm suggesting changing that directory tree to mention but not enumerate the common docs, and enumerate the track specific docs, with links to the common implementing doc(s) for specifics on the common docs.

And to give one counter-case for my suggestion: it could cause the document to be harder to use as it causes more 'flipping' back and forth between different documents to get a whole picture of what needs to be implemented.

**Comment by ErikSchierboom** [comment]
> But should we go a little further and remove the non-track-specific files from all the track implementing docs? 

@verdammelt I understand where this is coming from, but this _is_ an area where I don't expect many changes to occur. It is probably the most stable bit of the spec :) And as you said, there _is_ benefit in listing all files there. I'd opt for keeping all the files in the track listing.

File: docs/maintainers/generic-how-to-implement-a-concept-exercise.md [hunk 5]
```diff
@@ -28,41 +29,67 @@ languages
...
             └── &lt;SLUG&gt;
                 ├── .docs
+                |   ├── after.md
+                |   ├── hints.md
                 |   ├── instructions.md
                 |   ├── introduction.md
-                |   ├── hints.md
-                |   ├── source.md (required if there are third-party sources)
-                |   └── after.md
+                |   └── source.md (required if there are third-party sources)
```
Comment (SaschaMann) [inline docs/maintainers/generic-how-to-implement-a-concept-exercise.md:36]: The order doesn't match the track-specific files. The latter were reordered, so for consistency they should match.

File: docs/maintainers/generic-how-to-implement-a-concept-exercise.md [hunk 6]
```diff
@@ -28,41 +29,67 @@ languages
...
                     ├── config.json
                     └── design.md
 </pre>
 
-## Step 1: add .docs/introduction.md
+All Markdown files should adhere to the [style guide][style-guide], noting the [automatic formatting section][style-guide-auto-formatting]. Also check any language-specific style guides, where applicable.
+
+The following files need to be added and updated:
+
+## Add`.docs/introduction.md` file
```
Comment (SaschaMann) [inline docs/maintainers/generic-how-to-implement-a-concept-exercise.md:46]:
```suggestion
## Add `.docs/introduction.md` file
```

Comment (SaschaMann) [reply docs/maintainers/generic-how-to-implement-a-concept-exercise.md:46]: This appears to be wrong everywhere below, too

File: docs/maintainers/generic-how-to-implement-a-concept-exercise.md (hunk 5 above)
Comment (ErikSchierboom) [reply docs/maintainers/generic-how-to-implement-a-concept-exercise.md]: Done!

**Comment by ErikSchierboom** [comment]
Thanks everyone!

**Pull request merged**